datavault/backend.py

```diff
@@ -216,7 +439,379 @@ def hasMore(self, pos):
+        """Add a comment to the dataset."""
+        t = time.time()
+        new_comment = np.array([(t, user, comment)], dtype=self.comment_type)
+        old_comments = self.dataset.attrs['Comments']
+        data = np.hstack((old_comments, new_comment))
+        self.dataset.attrs.create('Comments', data, dtype=self.comment_type)
+
+    def getComments(self, limit, start):
+        """Get comments in [(datetime, username, comment), ...] format."""
+        if limit is None:
+            raw_comments = self.dataset.attrs['Comments'][start:]
+        else:
+            raw_comments = self.dataset.attrs['Comments'][start:start+limit]
+        comments = [(datetime.datetime.fromtimestamp(c[0]), str(c[1]), str(c[2])) for c in raw_comments]
+        return comments, start+len(comments)
+
+    def numComments(self):
+        return len(self.dataset.attrs['Comments'])
+
+class ExtendedHDF5Data(HDF5MetaData):
+    """Dataset backed by HDF5 file
+
+    This supports the extended dataset format which allows each column
+    to have a different type and to be arrays themselves.
+    """
+
+    def __init__(self, fh):
+        self._file = fh
+        if 'Version' not in self.file.attrs:
+            self.file.attrs['Version'] = np.asarray([3, 0, 0], dtype=np.int32)
+        self.version = np.asarray(self.file.attrs['Version'], np.int32)
+
+    def initialize_info(self, title, indep, dep):
+        """Initialize the columns when creating a new dataset"""
+        dtype = []
+        for idx, col in enumerate(indep + dep):
+            shape = col.shape
+            ttag = col.datatype
+            unit = col.unit
+            if len(shape) == 1 and shape[0] == 1:
+                shapestr = ''
+            else:
+                shapestr = str(tuple(shape))
+            varname = 'f{}'.format(idx)
+            if unit != '' and ttag not in ['v', 'c']:
+                raise RuntimeError('Unit {} specfied for datatype {}.  Only v and c may have units'.format(unit, ttag))
+            if ttag == 'i':
+                dtype.append((varname, shapestr + 'i4'))
+            elif ttag == 's':
+                if shapestr:
+                    raise ValueError("Cannot create string array column")
+                dtype.append((varname, h5py.special_dtype(vlen=str)))
+            elif ttag == 't':
+                dtype.append((varname, shapestr + 'i8'))
+            elif ttag == 'v':
+                dtype.append((varname, shapestr + 'f8'))
+            elif ttag == 'c':
+                dtype.append((varname, shapestr + 'c16'))
+            else:
+                raise RuntimeError("Invalid type tag {}".format(ttag))
+
+        self.file.create_dataset('DataVault', (0,), dtype=dtype, maxshape=(None,))
+        HDF5MetaData.initialize_info(self, title, indep, dep)
+
+    @property
+    def file(self):
+        return self._file()
+
+    @property
+    def dataset(self):
+        return self.file["DataVault"]
+
+    def addData(self, data):
```

Comment:
I think the number of columns will get checked by the underlying hdf5 library, since we're appending to a typed array. We should have tests for this, though.